applications/finetune-quora-embeddings/optimize.py

```diff
@@ -0,0 +1,266 @@
+    train_examples = format_dataset(train_dataset)
+    test_examples = format_dataset(test_dataset)
+    train_dataloader = DataLoader(
+        train_examples, shuffle=True, batch_size=params["batch_size"]
+    )
+
+    evaluator = evaluation.BinaryClassificationEvaluator.from_input_examples(
+        test_examples, batch_size=params["batch_size"]
+    )
+
+    model.fit(
+        optimizer_params={
+            "lr": params["learning_rate"],
+        },
+        train_objectives=[(train_dataloader, train_loss)],
+        evaluator=evaluator,
+        epochs=params["epochs"],
+        warmup_steps=params["warmup_steps"],
+        save_best_model=True,
+        show_progress_bar=True,
+        output_path=MODEL_SAVE_PATH,
+        scheduler=params["scheduler"],
+    )
+
+    model = SentenceTransformer(MODEL_SAVE_PATH)
+
+    predictions, test_labels = score_prediction(model, train_dataset, test_dataset)
+
+    eval_results = {
+        metric: function(test_labels, predictions)
+        for metric, function in METRICS.items()
+    }
+
+    existing_experiments.append(eval_results | params)
+    print("Logging results so far")
+    for idx, experiment in enumerate(existing_experiments):
+        print(f"Idx: {idx}, Result: {experiment}")
+
+    return OPTIMIZATION_METRIC_FUNC(test_labels, predictions)
+
+
+@stub.function(
+    image=image,
+    gpu=gpu_config,
+    network_file_systems={"/root/cache": STUDY_NFS},
+    volumes={DATASET_DIR: DATASET_VOLUME},
+    timeout=86400,
+)
+def optimize_hyperparameters(n_trials: int):
+    import optuna
+
+    storage = optuna.storages.JournalStorage(
+        optuna.storages.JournalFileStorage(JOURNAL_PATH)
+    )
+    results = []
+    # Set it so that we can resume a study in the event that something fails
+    study = optuna.create_study(
+        study_name="finetuning", storage=storage, load_if_exists=True
+    )
+
+    study.optimize(lambda trial: objective(trial, results), n_trials=n_trials)
+
+    return results
+
+
+@stub.function(
+    image=image,
+    network_file_systems={"/root/cache": STUDY_NFS},
+)
+def get_dataframe():
+    import optuna
+
+    storage = optuna.storages.JournalStorage(
+        optuna.storages.JournalFileStorage(JOURNAL_PATH)
+    )
+
+    # Set it so that we can resume a study in the event that something fails
+    study = optuna.create_study(
+        study_name="finetuning", storage=storage, load_if_exists=True
+    )
+
+    return study.trials_dataframe()
+
+
+@stub.local_entrypoint()
+def main():
+    import os
+    import csv
+
+    results = []
+    trials_per_worker = math.ceil(NUM_TRIALS / WORKERS)
+
+    for resp in optimize_hyperparameters.map(
+        [trials_per_worker for _ in range(WORKERS)],
+        order_outputs=False,
+        return_exceptions=True,
+    ):
+        if isinstance(resp, Exception):
+            print(f"Encountered Exception of {resp}")
+            continue
+
+        results.extend(resp)
+
+    # Check if the file exists
```

Comment:
seperate saving data code, and just use pandas, dont need to use csv.DictWriter. 

pd.Dataframes(**dict) (or records)